matching-core/src/main/java/com/graphhopper/matching/MapMatching.java

```diff
@@ -301,70 +328,92 @@ private void computeTransitionProbabilities(TimeStep<GPXExtension, GPXEntry, Pat
                     final double transitionLogProbability = probabilities
                             .transitionLogProbability(path.getDistance(), linearDistance, timeDiff);
                     timeStep.addTransitionLogProbability(from, to, transitionLogProbability);
+                } else {
+                    // TODO: can we remove maxVisitedNodes completely and just set to infinity?
+                    if (algo.getVisitedNodes() > algoOptions.getMaxVisitedNodes()) {
+                        throw new RuntimeException(
+                                "couldn't compute transition probabilities as routing failed due to too small maxVisitedNodes ("
+                                        + algoOptions.getMaxVisitedNodes() + ")");
+                    }
+                    // TODO: can we somewhere record that this route failed? Currently all viterbi
+                    // knows is that there's no transition possible, but not why. This is useful
+                    // for e.g. explaining why a sequence broke.
                 }
             }
         }
     }
 
-    private MatchResult computeMatchResult(List<SequenceState<GPXExtension, GPXEntry, Path>> seq,
-                                           List<GPXEntry> gpxList, List<QueryResult> allCandidates,
-                                           EdgeExplorer explorer) {
+    private MatchResult computeMatchResult(
+            List<List<SequenceState<GPXExtension, GPXEntry, Path>>> sequences,
+            List<GPXEntry> gpxList, List<QueryResult> allCandidates, EdgeExplorer explorer) {
         // every virtual edge maps to its real edge where the orientation is already correct!
         // TODO use traversal key instead of string!
         final Map<String, EdgeIteratorState> virtualEdgesMap = new HashMap<>();
         for (QueryResult candidate : allCandidates) {
             fillVirtualEdges(virtualEdgesMap, explorer, candidate);
         }
-
-        MatchResult matchResult = computeMatchedEdges(seq, virtualEdgesMap);
+        // TODO: sequences may be only single timesteps, or disconnected. So we need to support:
+        //  - fake edges e.g. 'got from X to Y' but we don't know how ...
+        //  - single points e.g. 'was at X from t0 to t1'
+        MatchResult matchResult = computeMatchedEdges(sequences, virtualEdgesMap);
         computeGpxStats(gpxList, matchResult);
 
         return matchResult;
     }
 
-    private MatchResult computeMatchedEdges(List<SequenceState<GPXExtension, GPXEntry, Path>> seq,
-                                            Map<String, EdgeIteratorState> virtualEdgesMap) {
-        List<EdgeMatch> edgeMatches = new ArrayList<>();
+    private MatchResult computeMatchedEdges(
+            List<List<SequenceState<GPXExtension, GPXEntry, Path>>> sequences,
+            Map<String, EdgeIteratorState> virtualEdgesMap) {
+        // TODO: remove gpx extensions and just add time at start/end of edge.
```

Comment:
Oh, that should (hopefully) be easy - I need to find the matching GPX entry to infer the start/end time anyway. Separate PR?